yt_dlp/extractor/msn.py

```diff
@@ -1,167 +1,208 @@
+            'duration': ('videoMetadata', 'playTime', {int_or_none}),
+            'tags': ('keywords', ..., {str}),
+            'uploader': ('provider', 'name', {str}),
+            'uploader_id': ('provider', 'id', {str}),
+        })
 
+        page_type = traverse_obj(json_data, ('type', {str}))
+        source_url = traverse_obj(json_data, ('sourceHref', {url_or_none}))
+        if page_type == 'video':
+            if traverse_obj(json_data, ('thirdPartyVideoPlayer', 'enabled')) and source_url:
+                return self.url_result(source_url)
             formats = []
-            for file_ in video.get('videoFiles', []):
-                format_url = file_.get('url')
-                if not format_url:
-                    continue
-                if 'format=m3u8-aapl' in format_url:
-                    # m3u8_native should not be used here until
-                    # https://github.com/ytdl-org/youtube-dl/issues/9913 is fixed
-                    formats.extend(self._extract_m3u8_formats(
-                        format_url, display_id, 'mp4',
-                        m3u8_id='hls', fatal=False))
-                elif 'format=mpd-time-csf' in format_url:
-                    formats.extend(self._extract_mpd_formats(
-                        format_url, display_id, 'dash', fatal=False))
-                elif '.ism' in format_url:
-                    if format_url.endswith('.ism'):
-                        format_url += '/manifest'
-                    formats.extend(self._extract_ism_formats(
-                        format_url, display_id, 'mss', fatal=False))
-                else:
-                    format_id = file_.get('formatCode')
-                    formats.append({
-                        'url': format_url,
-                        'ext': 'mp4',
-                        'format_id': format_id,
-                        'width': int_or_none(file_.get('width')),
-                        'height': int_or_none(file_.get('height')),
-                        'vbr': int_or_none(self._search_regex(r'_(\d+)\.mp4', format_url, 'vbr', default=None)),
-                        'quality': 1 if format_id == '1001' else None,
-                    })
-
             subtitles = {}
-            for file_ in video.get('files', []):
-                format_url = file_.get('url')
-                format_code = file_.get('formatCode')
-                if not format_url or not format_code:
-                    continue
-                if str(format_code) == '3100':
-                    subtitles.setdefault(file_.get('culture', 'en'), []).append({
-                        'ext': determine_ext(format_url, 'ttml'),
-                        'url': format_url,
-                    })
+            for file in traverse_obj(json_data, ('videoMetadata', 'externalVideoFiles', lambda _, v: url_or_none(v['url']))):
+                url = file['url']
+                ext = determine_ext(url)
```

Comment:
I don't like redefining `url`, doing so hurts maintainability and readability
```suggestion
                file_url = file['url']
                ext = determine_ext(file_url)
```